### PR TITLE
Add Github Enterprise support

### LINKED ISSUE #15: Support for Github Enterprise
Needed for https://github.com/download-directory/download-directory.github.io/issues/28

### PR DESCRIPTION
**Pull request by aslafy-z**
Fixes #15

### CODE REVIEW

**Comment by fregante** [comment]
I’m thinking that it should just accept a URL and we can parse all the data from it: api, user, repo, directory

Would you be willing to change that?

You can drop the previous user repo and directory parameters and I’ll release a new major version

**Comment by aslafy-z** [comment]
Would you expect the user to provide the ref or also infer it from that URL?
I'd suggest to keep the current implementation. It is easier to use for third party use-cases from my point of view.

**Comment by fregante** [comment]
Yeah, everything this module needs is in the URL, in fact it starts from a URL upstream and that tool has to include a GitHub URL parser and API endpoint selection logic. All of that should be in here. This is a wrapper around the API, you shouldn’t have to provide the API itself 

If we really want to preserve that flexibility, the first parameter of this module/function would be `resource`, which can be either a string (the URL) or an object with the 3-5 parts needed. Then it can go back to accepting 2 parameters: resource, options (token)

**Comment by aslafy-z** [comment]
I made the changes, but it still need some validation, and it's untested.

File: index.js [hunk 1]
```diff
@@ -1,33 +1,59 @@
...
+const urlParserRegex = /^[/]([^/]+)[/]([^/]+)[/]tree[/]([^/]+)[/](.*)/;
+
+function parseResource(res) {
+	if (typeof res === 'string') {
+		try {
+			const parsedUrl = new URL(res);
+			res = {api: 'https://api.github.com'};
+			[, res.user, res.repository, res.ref, res.directory] = urlParserRegex.exec(parsedUrl.pathname);
+			if (parsedUrl.hostname !== 'github.com') {
+				res.api = `${parsedUrl.protocol}//${parsedUrl.host}/api/v3`;
```
Comment (fregante) [inline index.js]: The protocol is always `https`
```suggestion
				res.api = `https://${parsedUrl.host}/api/v3`;
```

File: index.js [hunk 2]
```diff
@@ -1,33 +1,59 @@
...
+// Matches '/<re/po>/tree/<ref>/<dir>'
+const urlParserRegex = /^[/]([^/]+)[/]([^/]+)[/]tree[/]([^/]+)[/](.*)/;
+
+function parseResource(res) {
+	if (typeof res === 'string') {
+		try {
+			const parsedUrl = new URL(res);
+			res = {api: 'https://api.github.com'};
+			[, res.user, res.repository, res.ref, res.directory] = urlParserRegex.exec(parsedUrl.pathname);
+			if (parsedUrl.hostname !== 'github.com') {
```
Comment (fregante) [inline index.js]: I'd rather see a clearer:
```suggestion
			if (parsedUrl.hostname === 'github.com') {
				res.api = 'https://api.github.com';
			} else {
```

File: index.js [hunk 3]
```diff
@@ -1,33 +1,59 @@
...
-	const response = await fetch(`https://api.github.com/repos/${endpoint}${token}`);
+// Matches '/<re/po>/tree/<ref>/<dir>'
+const urlParserRegex = /^[/]([^/]+)[/]([^/]+)[/]tree[/]([^/]+)[/](.*)/;
+
+function parseResource(res) {
+	if (typeof res === 'string') {
+		try {
+			const parsedUrl = new URL(res);
+			res = {api: 'https://api.github.com'};
+			[, res.user, res.repository, res.ref, res.directory] = urlParserRegex.exec(parsedUrl.pathname);
```
Comment (fregante) [inline index.js]: This way you can avoid the `try`/`catch`

```suggestion
			[, res.user, res.repository, res.ref, res.directory] = urlParserRegex.exec(parsedUrl.pathname) || [];
			if (res.directory !== 'string') {
				 throw new Error('Unable to parse GitHub URL'); 
			}
```

File: index.js [hunk 4]
```diff
@@ -1,33 +1,59 @@
 const fetch = require('node-fetch'); // Automatically excluded in browser bundles
 
-async function api(endpoint, token) {
-	token = token ? `&access_token=${token}` : '';
-	const response = await fetch(`https://api.github.com/repos/${endpoint}${token}`);
+// Matches '/<re/po>/tree/<ref>/<dir>'
```
Comment (fregante) [inline index.js]:
```suggestion
// Matches '/<user>/<repo>/tree/<ref>/<dir>'
```

File: index.js [hunk 5]
```diff
@@ -1,33 +1,59 @@
...
 async function viaContentsApi({
-	user,
-	repository,
-	ref = 'HEAD',
-	directory,
+	resource,
 	token,
 	getFullData = false
 }) {
+	const res = parseResource(resource);
```
Comment (fregante) [inline index.js]: I think you can reuse it:
```suggestion
	resource = parseResource(resource);
```
Perhaps you can actually only parse the string:

```suggestion
	if (typeof resource === 'string') {
		resource = parseResource(resource);
	}
```

File: index.js [hunk 6]
```diff
@@ -41,21 +67,20 @@ async function viaContentsApi({
...
+	resource,
 	token,
 	getFullData = false
 }) {
-	if (!directory.endsWith('/')) {
-		directory += '/';
+	const res = parseResource(resource);
+
+	if (!res.directory.endsWith('/')) {
+		resource.directory += '/';
```
Comment (fregante) [inline index.js]: I think I had made a mistake here, it should be moved after the API call, or else the API endpoint becomes `user/repo//git/trees/ref?recursive=1`, notice the double `/` after repo

**Review by fregante — CHANGES_REQUESTED** [review]
Thanks for the changes! You can leave the validation out of this PR for now, it wasn't there before.

File: index.js (hunk 1 above)
Comment (fregante) [reply index.js]: This applies to everywhere you used this, including https://github.com/download-directory/download-directory.github.io/pull/29

File: index.js (hunk 6 above)
Comment (aslafy-z) [reply index.js]: Fixed!

**Comment by aslafy-z** [comment]
Not sure about that lint error:
```
  ✖  11:3  new Error() is too unspecific for a type check. Use new TypeError() instead.  unicorn/prefer-type-error
```

I disabled that line from now.

File: index.js [hunk 7]
```diff
@@ -1,33 +1,57 @@
...
+// Matches '/<user>/<repo>/tree/<ref>/<dir>'
+const urlParserRegex = /^[/]([^/]+)[/]([^/]+)[/]tree[/]([^/]+)[/](.*)/;
+
+function parseGithubURL(url) {
+	const res = {};
+	const parsedUrl = new URL(url);
+	[, res.user, res.repository, res.ref, res.directory] = urlParserRegex.exec(parsedUrl.pathname) || [];
+	if (typeof res.directory !== 'string') {
+		// eslint-disable-next-line unicorn/prefer-type-error
+		throw new Error('Unable to parse GitHub URL');
```
Comment (fregante) [inline index.js]: This fixes the lint

```suggestion
		throw new TypeError('Unable to parse GitHub URL');
```

File: index.js [hunk 8]
```diff
@@ -1,33 +1,57 @@
...
 }
 
 // Great for downloads with few sub directories on big repos
 // Cons: many requests if the repo has a lot of nested dirs
 async function viaContentsApi({
-	user,
-	repository,
-	ref = 'HEAD',
-	directory,
+	resource: res,
```
Comment (fregante) [inline index.js]: I think this could just be called `resource`?

```suggestion
	resource,
```

**Review by fregante — COMMENTED** [review]
Other than this it looks good!

Can you also extend this to test URLs as well? (of course not GHE)
https://github.com/fregante/list-github-dir-content/blob/master/tests.js

File: index.js (hunk 7 above)
Comment (aslafy-z) [reply index.js]: I wasn't sure whether it was correct to call it a TypeError. Ok for me.

File: index.js (hunk 8 above)
Comment (aslafy-z) [reply index.js]: It makes the references to that object below much more uselessly long. So I preferred to alias it. Let me know if you want me to change it.

**Comment by aslafy-z** [comment]
I had to revert some changes because API url was missing in case an object was passed without API in it, for it to default to Github public API, and ref needed default value too.

File: index.js (hunk 8 above)
Comment (fregante) [reply index.js]: It's `resource` in the first place, we have two names for the same thing. Also this rule will be enabled soon: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prevent-abbreviations.md

**Comment by fregante** [comment]
I published this as a prerelease. You can plug it into `download-directory` so we can test it live

```sh
npm install list-github-dir-content@3.0.0-0
```

**Review by fregante — APPROVED** [review]

File: index.js (hunk 8 above)
Comment (aslafy-z) [reply index.js]: Fixed!

**Review by fregante — CHANGES_REQUESTED** [review]
Can you review this? My tests seem to work, but I don't have GHE.

Also the readme needs to be updated, all the properties should be moved under the single `resource` item

**Comment by aslafy-z** [comment]
I'd love to see this working. I missed your message 3y ago and was pretty sure the PR was merged. Are you ok to merge an implementation of this feature today?

**Pull request closed without merging**